src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.h

```diff
@@ -372,6 +422,51 @@ class dfcc_instrumentt
     goto_programt::targett &target,
     goto_programt &goto_program,
     cfg_infot &cfg_info);
+
+  /// Builds a dependency graph representing the loop nesting structure.
+  void compute_loop_nesting_graph(
+    const irep_idt &function_id,
+    grapht<loop_graph_nodet> &loop_nesting_graph,
+    goto_programt &function_body);
+
+  /// Tags program instructions with pragmas
+  void tag_loop_instruction(
+    goto_programt::targett &target,
+    const irep_idt &pragma,
+    const size_t loop_id);
+
+  /// Return loop id pragma of a target.
+  irep_idt get_loop_id_pragma(const goto_programt::targett &target);
+
+  size_t
+  get_loop_id_from_target(const goto_programt::instructiont::targett target);
+
+  bool is_loop_instruction(const goto_programt::instructiont::targett target);
+
+  /// Remove loop id pragma from a target.
+  void remove_loop_id_pragma(goto_programt::targett &target);
+
+  // Infer assigns clauses for function with nested-graph-id `idx`.
+  void get_assigns_dfcc(
+    const local_may_aliast &local_may_alias,
+    goto_functionst::goto_functiont &goto_function,
+    const size_t idx,
+    assignst &assigns);
+
+  /// Add instructions for loop contracts.
+  void add_loop_contracts_instructions(
```

Comment:
could we move this to a separate class and split it in `add_pre_header_instuctions` `add_step_instructions`, `add_body_instructions`, `add_exit_instructions` ?